File: Trabalho 1/my_stack.py

class Node():
    def __init__(self, item):
        self.item = item
        self.next = None

class Stack():
    def __init__(self):
        
        self.head: Node | None = None
        
    def push(self, item):
        
        node = Node(item)
        
        if self.head == None:
            self.head = node
            return
        
        bottom = self.head
        self.head = node
        self.head.next = bottom
        return
        
    def pop(self):
        popped = self.head
        self.head = self.head.next
        return popped.item
    
    def isEmpty(self):
        if self.head == None:
            return True
        return False 
        
    def length(self):
        n = 0
        
        if self.isEmpty():
            return n
        
        actual = self.head
        
        while actual != None:
            n += 1
            
            actual = actual.next
            
        return n
    
    def top(self):
        return self.head.item
    
    
if __name__ == "__main__":
    stack1 = Stack()
    stack1.push(1)
    stack1.push(2)
    print(stack1.pop())
    print(stack1.isEmpty())
    print(stack1.length())
    print(stack1.top())

File: Trabalho 1 (Implementação em C)/DataStructures/my_stack.c
#include <stdio.h>
#include <stdlib.h>

#include "my_stack.h"
#include "graph_utils.h"

void Stack_init(Stack *stack) {
    stack->head = NULL;
}

void Stack_push(Stack *stack, int item) {
    StackNode *node = xmalloc(sizeof(StackNode));
    node->item = item;
    node->next = NULL;

    if (stack->head == NULL) {
        stack->head = node;
        return;
    }

    StackNode *bottom = stack->head;
    stack->head = node;
    stack->head->next = bottom;
}

int Stack_pop(Stack *stack) {
    StackNode *popped = stack->head;
    stack->head = stack->head->next;
    int item = popped->item;
    free(popped); /* em C o no removido precisa ser liberado */
    return item;
}

bool Stack_isEmpty(Stack *stack) {
    return stack->head == NULL;
}

int Stack_length(Stack *stack) {
    int n = 0;

    if (Stack_isEmpty(stack)) {
        return n;
    }

    StackNode *actual = stack->head;

    while (actual != NULL) {
        n += 1;
        actual = actual->next;
    }

    return n;
}

int Stack_top(Stack *stack) {
    return stack->head->item;
}

#ifdef MY_STACK_MAIN
/* if __name__ == "__main__" do my_stack.py */
int main(void) {
    Stack stack1;
    Stack_init(&stack1);
    Stack_push(&stack1, 1);
    Stack_push(&stack1, 2);
    printf("%d\n", Stack_pop(&stack1));
    printf("%s\n", Stack_isEmpty(&stack1) ? "True" : "False");
    printf("%d\n", Stack_length(&stack1));
    printf("%d\n", Stack_top(&stack1));
    return 0;
}
#endif
